### PR TITLE
Support Shutdown Gracefully

### PR DESCRIPTION
**Pull request by AlbumenJ**
## What is the purpose of the change

1. Add `unregister` to `Exporter`
2. `unregister` first when ServiceConfig `unexport`
3. wait for service idle in ServiceConfig
4. `offline` first in `ApplicationModelDeployer` and `ModuleModelDeployer`

## Checklist
- [x] Make sure there is a [GitHub_issue](https://github.com/apache/dubbo/issues) field for the change (usually before you start working on it). Trivial changes like typos do not require a GitHub issue. Your pull request should address just this issue, without pulling in other changes - one PR resolves one issue.
- [ ] Each commit in the pull request should have a meaningful subject line and body.
- [ ] Write a pull request description that is detailed enough to understand what the pull request does, how, and why.
- [ ] Check if is necessary to patch to Dubbo 3 if you are work on Dubbo 2.7
- [ ] Write necessary unit-test to verify your logic correction, more mock a little better when cross module dependency exist. If the new feature or significant change is committed, please remember to add sample in [dubbo samples](https://github.com/apache/dubbo-samples) project.
- [ ] Add some description to [dubbo-website](https://github.com/apache/dubbo-website) project if you are requesting to add a feature.
- [ ] GitHub Actions works fine on your own branch.
- [ ] If this contribution is large, please follow the [Software Donation Guide](https://github.com/apache/dubbo/wiki/Software-donation-guide).

### CODE REVIEW

File: dubbo-config/dubbo-config-api/src/main/java/org/apache/dubbo/config/ServiceConfig.java [hunk 1]
```diff
@@ -209,6 +219,51 @@ public void unexport() {
         repository.unregisterProvider(providerModel);
     }
 
+    private void waitForIdle() {
+        int timeout = ConfigurationUtils.getServerShutdownTimeout(getScopeModel());
+
+        long idleTime = System.currentTimeMillis() - providerModel.getLastInvokeTime();
```
Comment (QuentinYo) [inline dubbo-config/dubbo-config-api/src/main/java/org/apache/dubbo/config/ServiceConfig.java:225]: 为什么要根据providerModel.getLastInvokeTime()来计算idleTime呢，应该从unregister完成后开始计算更好吧

File: dubbo-config/dubbo-config-api/src/main/java/org/apache/dubbo/config/ServiceConfig.java [hunk 2]
```diff
@@ -209,6 +219,51 @@ public void unexport() {
...
+            } catch (InterruptedException e) {
+                logger.warn(INTERNAL_ERROR, "unknown error in registry module", "", e.getMessage(), e);
+                Thread.currentThread().interrupt();
+            }
+            return;
+        }
+
+        // 3. Wait for 3.33s(1/3 of shutdown time), if service has idle for 3.33s(1/3 of shutdown time), un-export directly,
+        //    otherwise wait for the rest time until idle for 3.33s(1/3 of shutdown time). The max wait time is 10s(shutdown time).
+        idleTime = 0;
```
Comment (QuentinYo) [inline dubbo-config/dubbo-config-api/src/main/java/org/apache/dubbo/config/ServiceConfig.java:247]: 可能最少只等待3.33s？

File: dubbo-config/dubbo-config-api/src/main/java/org/apache/dubbo/config/ServiceConfig.java (hunk 1 above)
Comment (AlbumenJ) [reply dubbo-config/dubbo-config-api/src/main/java/org/apache/dubbo/config/ServiceConfig.java:225]: 如果已经 idle 了 10s，认为没有请求了

File: dubbo-config/dubbo-config-api/src/main/java/org/apache/dubbo/config/ServiceConfig.java (hunk 2 above)
Comment (AlbumenJ) [reply dubbo-config/dubbo-config-api/src/main/java/org/apache/dubbo/config/ServiceConfig.java:247]: 是的，这里有个时间窗口的决策，时间窗口太小了可能导致推送没到位就关闭了

**Comment by QuentinYo** [comment]
请问有计划对3.0.x和3.1.x修复这个问题吗 https://github.com/apache/dubbo/issues/12064

File: dubbo-config/dubbo-config-api/src/main/java/org/apache/dubbo/config/ServiceConfig.java (hunk 1 above)
Comment (QuentinYo) [reply dubbo-config/dubbo-config-api/src/main/java/org/apache/dubbo/config/ServiceConfig.java:225]: 想一个极端场景：业务服务就是每10s调用一次，会不会在某次请求达到一个正在停机的provider上却处理不了呢。

File: dubbo-config/dubbo-config-api/src/main/java/org/apache/dubbo/config/ServiceConfig.java [hunk 3]
```diff
@@ -209,6 +219,51 @@ public void unexport() {
         repository.unregisterProvider(providerModel);
     }
 
+    private void waitForIdle() {
```
Comment (CrazyHZM) [inline dubbo-config/dubbo-config-api/src/main/java/org/apache/dubbo/config/ServiceConfig.java:222]: Graceful offline also needs to consider that the connection is no longer receiving new requests, and the thread pool has completed all tasks within the specified time. Is it possible to check the number of tasks in some thread pools here?

Comment (AlbumenJ) [reply dubbo-config/dubbo-config-api/src/main/java/org/apache/dubbo/config/ServiceConfig.java:222]: Concurrent check can be finished seperately. This PR is focus on wait for consumer not to use the offlined provider.

**Comment by AlbumenJ** [comment]
> 请问有计划对3.0.x和3.1.x修复这个问题吗 #12064

目前暂无计划

File: dubbo-config/dubbo-config-api/src/main/java/org/apache/dubbo/config/ServiceConfig.java [hunk 4]
```diff
@@ -194,6 +196,14 @@ public void unexport() {
...
         }
         if (!exporters.isEmpty()) {
+            for (Exporter<?> exporter : exporters) {
+                try {
+                    exporter.unregister();
+                } catch (Throwable t) {
+                    logger.warn(CONFIG_UNEXPORT_ERROR, "", "", "Unexpected error occurred when unexport " + exporter, t);
+                }
+            }
+            waitForIdle();
```
Comment (QuentinYo) [inline dubbo-config/dubbo-config-api/src/main/java/org/apache/dubbo/config/ServiceConfig.java:206]: 如果一个dubbo项目有多个接口（ServiceBean），每个ServiceBean进行unexport的时候都需要等待，会导致停机时间变成10*n

Comment (AlbumenJ) [reply dubbo-config/dubbo-config-api/src/main/java/org/apache/dubbo/config/ServiceConfig.java:206]: 在 Module、Application 关闭的时候会提前 offline 所有服务。所以理论最长时间只有 10s

Comment (QuentinYo) [reply dubbo-config/dubbo-config-api/src/main/java/org/apache/dubbo/config/ServiceConfig.java:206]: offline 那部分逻辑貌似无法下掉应用级的注册信息。
org.apache.dubbo.registry.client.ServiceDiscoveryRegistry#unregister
image

org.apache.dubbo.registry.client.AbstractServiceDiscovery#unregister(org.apache.dubbo.common.URL)
image

Comment (QuentinYo) [reply dubbo-config/dubbo-config-api/src/main/java/org/apache/dubbo/config/ServiceConfig.java:206]: 应用级服务下线需要配合下面的配置。（默认配置最大30s后才同步到注册中心）
org.apache.dubbo.config.deploy.DefaultApplicationDeployer#registerServiceInstance
image

Comment (AlbumenJ) [reply dubbo-config/dubbo-config-api/src/main/java/org/apache/dubbo/config/ServiceConfig.java:206]: 3.2 目前默认 1s 了

**Pull request merged**